examples/piecestore/rpc/client/api/api.go

```diff
@@ -0,0 +1,122 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package api
+
+import (
+	"errors"
+	"io"
+	"log"
+
+	"github.com/zeebo/errs"
+
+	"golang.org/x/net/context"
+
+	pb "storj.io/storj/protos/piecestore"
+)
+
+var serverError = errs.Class("serverError")
+
+// PieceMeta -- Struct containing Piece information from PieceMetaRequest
+type PieceMeta struct {
+	ID         string
+	Size       int64
+	Expiration int64
+}
+
+// PieceMetaRequest -- Request info about a piece by Id
+func PieceMetaRequest(ctx context.Context, c pb.PieceStoreRoutesClient, id string) (*PieceMeta, error) {
+
+	reply, err := c.Piece(ctx, &pb.PieceId{Id: id})
+	if err != nil {
+		return nil, err
+	}
+
+	return &PieceMeta{ID: reply.Id, Size: reply.Size, Expiration: reply.Expiration}, nil
+}
+
+// StorePieceRequest -- Upload Piece to Server
+func StorePieceRequest(ctx context.Context, c pb.PieceStoreRoutesClient, id string, data io.Reader, dataOffset int64, length int64, ttl int64, storeOffset int64) error {
+	stream, err := c.Store(ctx)
+
+	// SSend preliminary data
+	if err := stream.Send(&pb.PieceStore{Id: id, Size: length, Ttl: ttl, StoreOffset: storeOffset}); err != nil {
+		return err
+	}
+
+	buffer := make([]byte, 4096)
```

Comment:
so believe it or not, context switches are significant enough that most performant copy loops have moved away from using the old hard drive block size as a buffer a while back. disks use larger buffers usually and so does the kernel. so even though 4096 has a lot of history, the go io.Copy implementation actually uses 32768 (32*1024).

one thing you could do to make it so you don't have to worry about buffer sizes or anything is make a small wrapper around the `stream` that makes it match the io.Writer interface (calls to `Write` just get passed to `stream.Send`). That way you could just use io.Copy directly and you wouldn't even need this for loop

Reply:
this is not a big enough issue to hold up merging but if you have time and another review cycle anyways, could be good

Reply:
👍

Reply:
👍